Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm

```diff
@@ -121,19 +173,40 @@
         parsedDetails.type = parsedType;
     }
 
-    MockBookmarkNode newNode;
-    newNode.id = bookmark[idKey];
-    newNode.url = bookmark[urlKey];
-    newNode.title = bookmark[titleKey];
-    newNode.parentId = bookmark[parentIdKey];
-    newNode.index = parsedDetails.index.value();
-    newNode.type = parsedDetails.type.value_or(WebExtensionAPIBookmarks::BookmarkTreeNodeType::Folder);
+    Ref newNode = MockBookmarkNode::create();
+    newNode->id = bookmark[idKey];
+    newNode->url = bookmark[urlKey];
+    newNode->title = bookmark[titleKey];
+    newNode->parentId = bookmark[parentIdKey] ? bookmark[parentIdKey] : bookmarksRootId;
+    newNode->index = parsedDetails.index.value();
+    newNode->type = bookmark[urlKey] ? WebExtensionAPIBookmarks::BookmarkTreeNodeType::Bookmark : WebExtensionAPIBookmarks::BookmarkTreeNodeType::Folder;
 
     double msSinceEpoch = dynamic_objc_cast<NSNumber>(bookmark[dateAddedKey]).doubleValue;
-    newNode.dateAdded = WTF::WallTime::fromRawSeconds(msSinceEpoch / 1000.0);
+    newNode->dateAdded = WTF::WallTime::fromRawSeconds(msSinceEpoch / 1000.0);
+
+    String parentId = newNode->parentId;
+    String newNodeId = newNode->id;
+
+    auto parentOptional = m_mockBookmarks.getOptional(parentId);
+    if (!parentOptional) {
+        *outExceptionString = toErrorString(nullString(), parentIdKey, @"it could not be mapped to a node").createNSString().autorelease();
+        return;
+    }
+    Ref parentNode = parentOptional.value();
 
-    m_mockBookmarks.append(WTFMove(newNode));
-    callback->call(toWebAPI(m_mockBookmarks.last()));
+    if (parentNode->type != WebExtensionAPIBookmarks::BookmarkTreeNodeType::Folder) {
+        *outExceptionString = toErrorString(nullString(), parentIdKey, @"it must specify a node which is a folder").createNSString().autorelease();
+        return;
+    }
+
+    m_mockBookmarks.add(newNodeId, newNode);
+
+    parentNode->children.append(newNode);
+    std::sort(parentNode->children.begin(), parentNode->children.end(),
+        [](const Ref<MockBookmarkNode>& a, const Ref<MockBookmarkNode>& b) {
+            return a->index < b->index;
+        });
```

Comment:
Nit: This is a bit of extra work. When the original parent folder was created, it was sorted, right (in `initializeMockBookmarksInternal`)? Shouldn't we be able to just insert in order here, instead of append and then sort?

However, given that this is just for mock bookmark nodes, I don't think you need to actually make a change here.